File: scratch/inspect_questions.js
const fs = require('fs');

const content = fs.readFileSync('E:/workspace/agy/AI_TEST/index.html', 'utf8');

const startMarker = 'const fullQuestions = [';
const startIdx = content.indexOf(startMarker);
if (startIdx === -1) {
    console.error("Could not find start marker!");
    process.exit(1);
}

// Track square bracket balance to find the end of the array
let balance = 0;
let arrayStart = startIdx + startMarker.length - 1; // index of '['
let endIdx = -1;

for (let i = arrayStart; i < content.length; i++) {
    if (content[i] === '[') balance++;
    else if (content[i] === ']') {
        balance--;
        if (balance === 0) {
            endIdx = i;
            break;
        }
    }
}

if (endIdx === -1) {
    console.error("Could not find end bracket!");
    process.exit(1);
}

const jsArrayText = content.slice(arrayStart, endIdx + 1);
console.log(`Extracted JS array text length: ${jsArrayText.length}`);

// We can evaluate the array using eval or Function inside Node.js!
let fullQuestions;
try {
    fullQuestions = eval(jsArrayText);
} catch (e) {
    console.error("Failed to parse array with eval: " + e.message);
    process.exit(1);
}

console.log(`Successfully parsed ${fullQuestions.length} questions from index.html!`);

const issues = [];
fullQuestions.forEach((q, idx) => {
    const qId = q.id || `UNKNOWN_ID_at_index_${idx}`;
    if (!q.id) issues.push(`Question at index ${idx}: missing 'id'`);
    if (!q.year) issues.push(`Q ${qId}: missing 'year'`);
    if (!q.num) issues.push(`Q ${qId}: missing 'num'`);
    if (!q.type) issues.push(`Q ${qId}: missing 'type'`);
    if (!q.question) issues.push(`Q ${qId}: missing 'question'`);
    if (!q.options) {
        issues.push(`Q ${qId}: missing 'options'`);
    } else {
        if (!Array.isArray(q.options)) {
            issues.push(`Q ${qId}: 'options' is not an array`);
        } else if (q.options.length !== 4) {
            issues.push(`Q ${qId}: 'options' has length ${q.options.length}, expected 4`);
        } else {
            q.options.forEach((opt, optIdx) => {
                if (typeof opt !== 'string' || opt.trim() === '') {
                    issues.push(`Q ${qId} option ${optIdx}: empty or non-string option`);
                }
            });
        }
    }
    if (!q.answer) {
        issues.push(`Q ${qId}: missing 'answer'`);
    } else {
        if (!['A', 'B', 'C', 'D'].includes(q.answer)) {
            issues.push(`Q ${qId}: 'answer' is '${q.answer}', expected A, B, C, or D`);
        }
    }
    if (!q.explanation || q.explanation.trim().length < 10) {
        issues.push(`Q ${qId}: 'explanation' is missing or too short`);
    }
});

if (issues.length > 0) {
    console.log(`Found ${issues.length} issues:`);
    issues.forEach(issue => console.log(' - ' + issue));
} else {
    console.log("ALL 150 QUESTIONS ARE 100% VALID AND CORRECTLY STRUCTURED!");
}

File: scratch/inspect_questions.py
with open("E:/workspace/agy/AI_TEST/index.html", "r", encoding="utf-8") as f:
    content = f.read()

start_marker = "const fullQuestions = ["
start_idx = content.find(start_marker)
if start_idx == -1:
    print("Could not find start marker!")
    exit(1)

# Find the end of the array by tracking the square bracket balance starting after "const fullQuestions = "
# The array starts with '['
balance = 0
array_start = start_idx + len("const fullQuestions = ") - 1 # this is the '['
end_idx = -1

for idx in range(array_start, len(content)):
    char = content[idx]
    if char == '[':
        balance += 1
    elif char == ']':
        balance -= 1
        if balance == 0:
            end_idx = idx
            break

if end_idx == -1:
    print("Could not find matching end square bracket!")
    exit(1)

js_array_text = content[array_start : end_idx + 1]
print(f"Extracted JS array length: {len(js_array_text)}")

# Let's parse all objects within this array by tracking brace balance
brace_count = 0
current_obj = ""
in_obj = False
questions = []

import ast

for char in js_array_text:
    if char == '{':
        if brace_count == 0:
            in_obj = True
            current_obj = ""
        brace_count += 1
    
    if in_obj:
        current_obj += char
        
    if char == '}':
        brace_count -= 1
        if brace_count == 0:
            in_obj = False
            # Clean and parse current_obj
            obj_str = current_obj.strip()
            obj_str_cleaned = obj_str.replace("true", "True").replace("false", "False").replace("null", "None")
            try:
                parsed = ast.literal_eval(obj_str_cleaned)
                questions.append(parsed)
            except Exception as e:
                # Let's print the error and a snippet if it fails
                print(f"Failed to parse object: {e}")
                print(obj_str[:150])
                print("...")

print(f"Total parsed questions: {len(questions)}")

# Let's print summary of questions per year
years = {}
for q in questions:
    y = q.get("year")
    years[y] = years.get(y, 0) + 1
print("Questions count per year:", years)

# Check for issues
issues = []
for q in questions:
    q_id = q.get("id")
    year = q.get("year")
    num = q.get("num")
    options = q.get("options", [])
    answer = q.get("answer")
    explanation = q.get("explanation")
    
    # 1. Check if options is a list of 4 elements
    if not isinstance(options, list) or len(options) != 4:
        issues.append(f"Q {q_id}: 'options' is not of length 4. Got {options}")
        
    # 2. Check if answer is A, B, C, or D
    if answer not in ["A", "B", "C", "D"]:
        issues.append(f"Q {q_id}: 'answer' is not A, B, C, or D. Got '{answer}'")
        
    # 3. Check if explanation exists and is detailed
    if not explanation or len(explanation.strip()) < 10:
        issues.append(f"Q {q_id}: 'explanation' is missing or too short. Got '{explanation}'")

if issues:
    print(f"Found {len(issues)} potential issues:")
    for issue in issues:
        print(" -", issue)
else:
    print("No issues found in the full 150 questions array!")
